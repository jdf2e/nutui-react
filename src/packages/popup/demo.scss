.demo-popup {
  color: $gray6;
}
