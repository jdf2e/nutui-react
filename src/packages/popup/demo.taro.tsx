import React, { useState } from 'react'
import { Failure, Heart } from '@nutui/icons-react-taro'
import Taro from '@tarojs/taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import { Cell, Popup } from '@/packages/nutui.react.taro'
import Header from '@/sites/components/header'

interface T {
  ce5c5446: string
  c38a08ef: string
  b840c88f: string
  a74a1fd4: string
  '8dab2f66': string
  cfbdc781: string
  c3a3a1d2: string
  e51e4582: string
  '7db1a8b2': string
  '7db1a8b3': string
  '7db1a8b4': string
  '7db1a8b5': string
  a52bef0c: string
  d04fcbda: string
  '0aaad620': string
  ea3d02f2: string
  c9e6df49: string
}

const PopupDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      ce5c5446: '基础类型',
      c38a08ef: '展示弹出层',
      b840c88f: '正文',
      a74a1fd4: '弹出位置',
      '8dab2f66': '顶部弹出',
      cfbdc781: '底部弹出',
      c3a3a1d2: '左侧弹出',
      e51e4582: '右侧弹出',
      '7db1a8b2': '关闭图标',
      '7db1a8b3': '阻塞关闭',
      '7db1a8b4': '阻塞点击 Overlay 关闭',
      '7db1a8b5': '阻塞点击 close icon 关闭',
      a52bef0c: '图标位置',
      d04fcbda: '自定义图标',
      '0aaad620': '圆角弹框',
      ea3d02f2: '指定节点挂载',
      c9e6df49: '多层堆叠',
    },
    'zh-TW': {
      ce5c5446: '基礎類型',
      c38a08ef: '展示彈出層',
      b840c88f: '正文',
      a74a1fd4: '彈出位置',
      '8dab2f66': '頂部彈出',
      cfbdc781: '底部彈出',
      c3a3a1d2: '左側彈出',
      e51e4582: '右側彈出',
      '7db1a8b2': '關閉圖標',
      '7db1a8b3': '阻塞關閉',
      '7db1a8b4': '阻塞点击 Overlay 关闭',
      '7db1a8b5': '阻塞点击 close icon 关闭',
      a52bef0c: '圖標位置',
      d04fcbda: '自定義圖標',
      '0aaad620': '圓角彈框',
      ea3d02f2: '指定節點掛載',
      c9e6df49: '多層堆疊',
    },
    'en-US': {
      ce5c5446: 'base type',
      c38a08ef: 'Show popup',
      b840c88f: 'text',
      a74a1fd4: 'popup location',
      '8dab2f66': 'top pop',
      cfbdc781: 'bottom pop',
      c3a3a1d2: 'pop up left',
      e51e4582: 'pop up right',
      '7db1a8b2': 'close icon',
      '7db1a8b3': 'stop close',
      '7db1a8b4': 'when click Overlay,stop close',
      '7db1a8b5': 'when click close icon, stop close',
      a52bef0c: 'Icon position',
      d04fcbda: 'custom icon',
      '0aaad620': 'Rounded popup',
      ea3d02f2: 'Mount the specified node',
      c9e6df49: 'multi-layer stacking',
    },
  })

  const [showBasic, setShowBasic] = useState(false)
  const [showTop, setShowTop] = useState(false)
  const [showBottom, setShowBottom] = useState(false)
  const [showLeft, setShowLeft] = useState(false)
  const [showRight, setShowRight] = useState(false)
  const [showIcon, setShowIcon] = useState(false)
  const [showIconPosition, setShowIconPosition] = useState(false)
  const [showIconDefine, setShowIconDefine] = useState(false)
  const [showBottomRound, setShowBottomRound] = useState(false)
  const [showMountNode, setShowMountNode] = useState(false)
  const [showMutiple, setShowMutiple] = useState(false)
  const [showMutipleInner, setShowMutipleInner] = useState(false)
  const [showOverlayStop, setShowOverlayStop] = useState(false)
  const [showCloseIconStop, setShowCloseIconStop] = useState(false)

  return (
    <>
      <Header />
      <div className={`demo ${Taro.getEnv() === 'WEB' ? 'web' : ''}`}>
        <h2>{translated.ce5c5446}</h2>
        <Cell
          title={translated.c38a08ef}
          onClick={() => {
            setShowBasic(true)
          }}
        />
        <Popup
          visible={showBasic}
          style={{ padding: '30px 50px' }}
          onClose={() => {
            setShowBasic(false)
          }}
        >
          {translated.b840c88f}
        </Popup>

        <h2>{translated.a74a1fd4}</h2>
        <Cell
          title={translated['8dab2f66']}
          onClick={() => {
            setShowTop(true)
          }}
        />
        <Popup
          visible={showTop}
          destroyOnClose
          position="top"
          onClose={() => {
            setShowTop(false)
          }}
        />
        <Cell
          title={translated.cfbdc781}
          onClick={() => {
            setShowBottom(true)
          }}
        />
        <Popup
          visible={showBottom}
          position="bottom"
          onClose={() => {
            setShowBottom(false)
          }}
        />
        <Cell
          title={translated.c3a3a1d2}
          onClick={() => {
            setShowLeft(true)
          }}
        />
        <Popup
          visible={showLeft}
          style={{ width: '40%', height: '100%' }}
          position="left"
          onClose={() => {
            setShowLeft(false)
          }}
        />
        <Cell
          title={translated.e51e4582}
          onClick={() => {
            setShowRight(true)
          }}
        />
        <Popup
          visible={showRight}
          style={{ width: '40%', height: '100%' }}
          position="right"
          onClose={() => {
            setShowRight(false)
          }}
        />

        <h2>{translated['7db1a8b2']}</h2>
        <Cell
          title={translated['7db1a8b2']}
          onClick={() => {
            setShowIcon(true)
          }}
        />
        <Popup
          closeable
          visible={showIcon}
          left="返回"
          title="我是标题"
          position="bottom"
          onClose={() => {
            setShowIcon(false)
          }}
        />
        <Cell
          title={translated.a52bef0c}
          onClick={() => {
            setShowIconPosition(true)
          }}
        />
        <Popup
          closeable
          visible={showIconPosition}
          closeIconPosition="top-left"
          position="bottom"
          onClose={() => {
            setShowIconPosition(false)
          }}
        />
        <Cell
          title={translated.d04fcbda}
          onClick={() => {
            setShowIconDefine(true)
          }}
        />
        <Popup
          closeable
          closeIcon={<Heart size="15px" />}
          visible={showIconDefine}
          style={{ height: '100%' }}
          position="bottom"
          onClose={() => {
            setShowIconDefine(false)
          }}
        />

        <h2>{translated[`7db1a8b3`]}</h2>
        <Cell
          title={translated[`7db1a8b4`]}
          onClick={() => {
            setShowOverlayStop(true)
          }}
        />
        <Popup
          visible={showOverlayStop}
          style={{ padding: '30px 50px' }}
          onClose={() => {
            setShowOverlayStop(false)
          }}
          onOverlayClick={() => {
            console.log('onOverlayClick')
            return false
          }}
        >
          {translated.b840c88f}
        </Popup>
        <Cell
          title={translated[`7db1a8b5`]}
          onClick={() => {
            setShowCloseIconStop(true)
          }}
        />
        <Popup
          closeable
          closeIcon={<Failure width="12px" height="12px" />}
          visible={showCloseIconStop}
          closeOnOverlayClick={false}
          style={{ height: '40%' }}
          position="bottom"
          onClose={() => {
            setShowIcon(false)
          }}
          onCloseIconClick={() => {
            console.log('onCloseIconClick')
          }}
        />

        <h2>{translated['0aaad620']}</h2>
        <Cell
          title={translated['0aaad620']}
          onClick={() => {
            setShowBottomRound(true)
          }}
        />
        <Popup
          visible={showBottomRound}
          style={{ height: '20%' }}
          position="top"
          round
          onClose={() => {
            setShowBottomRound(false)
          }}
        />
        {/* 等一下3.5.0 */}
        {/* <h2>{translated.ea3d02f2}</h2>
        <Cell
          title={translated.ea3d02f2}

          onClick={() => {
            setShowMountNode(true)
          }}
        />
        <Popup
          visible={showMountNode}
          style={{ padding: '30px 50px' }}
          portal={document.body}
          onClose={() => {
            setShowMountNode(false)
          }}
        >
          body
        </Popup> */}

        <h2>{translated.c9e6df49}</h2>
        <Cell
          title={translated.c9e6df49}
          onClick={() => {
            setShowMutiple(true)
          }}
        />
        <Popup
          visible={showMutiple}
          style={{ padding: '30px 50px' }}
          onClose={() => {
            setShowMutiple(false)
          }}
        >
          <span
            onClick={() => {
              setShowMutipleInner(true)
            }}
          >
            Click It
          </span>
        </Popup>
        <Popup
          visible={showMutipleInner}
          style={{ padding: '30px 50px' }}
          overlayStyle={{ backgroundColor: 'transparent' }}
          onClose={() => {
            setShowMutipleInner(false)
          }}
        >
          <span
            onClick={() => {
              setShowMutipleInner(false)
            }}
          >
            close
          </span>
        </Popup>
      </div>
    </>
  )
}

export default PopupDemo
