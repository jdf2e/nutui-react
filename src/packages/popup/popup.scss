@import '../overlay/overlay.scss';

.nut-popup {
  position: fixed;
  max-height: 100%;
  overflow-y: auto;
  background-color: $overlay-content-bg-color;
  -webkit-overflow-scrolling: touch;

  &-title {
    display: flex;
    align-items: center;
    justify-content: center;
    color: $title-color;
    padding: $popup-title-padding;
    font-weight: $font-weight-bold;
    line-height: $popup-title-line-height;
    font-size: $popup-title-font-size;
    border-bottom: $popup-title-border-bottom;
  }

  &-left-icon {
    position: absolute;
    top: $popup-close-icon-margin;
    left: $popup-close-icon-margin;
  }

  &__close-icon {
    position: absolute !important;
    z-index: 1;
    cursor: pointer;
    width: 30px;
    height: 30px;
    display: flex;
    justify-content: center;
    align-items: center;

    &:active {
      opacity: 0.7;
    }

    &--top-left {
      top: $popup-close-icon-margin;
      left: $popup-close-icon-margin;
    }

    &--top-right {
      top: $popup-close-icon-margin;
      right: $popup-close-icon-margin;
    }

    &--bottom-left {
      bottom: $popup-close-icon-margin;
      left: $popup-close-icon-margin;
    }

    &--bottom-right {
      right: $popup-close-icon-margin;
      bottom: $popup-close-icon-margin;
    }
  }
}

.nut-popup-center {
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);

  &.round {
    border-radius: $popup-border-radius;
  }
}

.nut-popup-bottom {
  bottom: 0;
  left: 0;
  width: 100%;

  &.round {
    border-radius: $popup-border-radius $popup-border-radius 0 0;
  }
}

.nut-popup-right {
  top: 0;
  right: 0;

  &.round {
    border-radius: $popup-border-radius 0 0 $popup-border-radius;
  }
}

.nut-popup-left {
  top: 0;
  left: 0;

  &.round {
    border-radius: 0 $popup-border-radius $popup-border-radius 0;
  }
}

.nut-popup-top {
  top: 0;
  left: 0;
  width: 100%;

  &.round {
    border-radius: 0 0 $popup-border-radius $popup-border-radius;
  }
}

.nut-popup-slide {
  @keyframes popup-scale-fade-in {
    from {
      opacity: 0;
      transform: scale(0.8);
    }

    to {
      opacity: 1;
      transform: scale(1);
    }
  }

  @keyframes popup-scale-fade-out {
    from {
      opacity: 1;
      transform: scale(1);
    }

    to {
      opacity: 0;
      transform: scale(0.8);
    }
  }

  &-default-enter-active {
    animation-fill-mode: both;
    animation-name: popup-scale-fade-in;
    animation-duration: $popup-animation-duration;
  }

  &-default-exit-active {
    animation-fill-mode: both;
    animation-name: popup-scale-fade-out;
    animation-duration: $popup-animation-duration;
  }

  @keyframes popup-fade-in {
    from {
      opacity: 0;
    }

    to {
      opacity: 1;
    }
  }

  @keyframes popup-fade-out {
    from {
      opacity: 1;
    }

    to {
      opacity: 0;
    }
  }

  &-center-enter-active {
    animation-fill-mode: both;
    animation-name: popup-fade-in;
    animation-duration: $popup-animation-duration;
  }

  &-center-exit-active {
    animation-fill-mode: both;
    animation-name: popup-fade-out;
    animation-duration: $popup-animation-duration;
  }

  /* 从顶部滑出 */
  @keyframes popup-slide-top-enter {
    from {
      transform: translate3d(0, -100%, 0);
    }
  }

  @keyframes popup-slide-top-exit {
    to {
      transform: translate3d(0, -100%, 0);
    }
  }

  &-top-enter-active,
  &-top-appear-active {
    transform: translate3d(0, 0%, 0);
    animation-fill-mode: both;
    animation-name: popup-slide-top-enter;
    animation-duration: $popup-animation-duration;
  }

  &-top-exit-active {
    animation-fill-mode: both;
    animation-name: popup-slide-top-exit;
    animation-duration: $popup-animation-duration;
  }

  /* 从右侧滑出 */
  @keyframes popup-slide-right-enter {
    from {
      transform: translate3d(100%, 0, 0);
    }
  }

  @keyframes popup-slide-right-exit {
    to {
      transform: translate3d(100%, 0, 0);
    }
  }

  &-right-enter-active,
  &-right-appear-active {
    transform: translate3d(0, 0, 0);
    animation-fill-mode: both;
    animation-name: popup-slide-right-enter;
    animation-duration: $popup-animation-duration;
  }

  &-right-exit {
    animation-fill-mode: both;
    animation-name: popup-slide-right-exit;
    animation-duration: $popup-animation-duration;
  }

  /* 从底部滑出 */
  @keyframes popup-slide-bottom-enter {
    from {
      transform: translate3d(0, 100%, 0);
    }
  }

  @keyframes slide-bottom-exit {
    to {
      transform: translate3d(0, 100%, 0);
    }
  }

  &-bottom-enter-active,
  &-bottom-appear-active {
    transform: translate(0, 0);
    animation-fill-mode: both;
    animation-name: popup-slide-bottom-enter;
    animation-duration: $popup-animation-duration;
  }

  &-bottom-exit {
    animation-fill-mode: both;
    animation-name: slide-bottom-exit;
    animation-duration: $popup-animation-duration;
  }

  /* 从左侧滑出 */
  @keyframes popup-slide-left-enter {
    from {
      transform: translate3d(-100%, 0, 0);
    }
  }

  @keyframes popup-slide-left-exit {
    to {
      transform: translate3d(-100%, 0, 0);
    }
  }

  &-left-enter-active,
  &-left-appear-active {
    transform: translate(0, 0);
    animation-fill-mode: both;
    animation-name: popup-slide-left-enter;
    animation-duration: $popup-animation-duration;
  }

  &-left-exit-active,
  &-left-exit-done {
    animation-fill-mode: both;
    animation-name: popup-slide-left-exit;
    animation-duration: $popup-animation-duration;
  }
}

.nut-overflow-hidden {
  overflow: hidden !important;
}
