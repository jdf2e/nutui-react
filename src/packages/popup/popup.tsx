import React, {
  FunctionComponent,
  useState,
  useEffect,
  MouseEventHandler,
  MouseEvent,
  ReactElement,
  ReactPortal,
  ReactNode,
} from 'react'
import { createPortal } from 'react-dom'
import { CSSTransition } from 'react-transition-group'
import classNames from 'classnames'
import { Close } from '@nutui/icons-react'
import { EnterHandler, ExitHandler } from 'react-transition-group/Transition'
import { OverlayProps, defaultOverlayProps } from '@/packages/overlay/overlay'
import Overlay from '@/packages/overlay'
import { ComponentDefaults } from '@/utils/typings'
import { useLockScroll } from '@/utils/use-lock-scroll'

type Teleport = HTMLElement | (() => HTMLElement) | null

export interface PopupProps extends OverlayProps {
  position: string
  transition: string
  overlayStyle: React.CSSProperties
  overlayClassName: string
  closeable: boolean
  closeIconPosition: string
  closeIcon: ReactNode
  left: ReactNode
  title: ReactNode
  destroyOnClose: boolean
  portal: Teleport
  overlay: boolean
  round: boolean
  onOpen: () => void
  onClose: () => void
  onOverlayClick: (e: MouseEvent) => boolean | void
  onCloseIconClick: (e: MouseEvent) => boolean | void
}

const defaultProps = {
  ...ComponentDefaults,
  position: 'center',
  transition: '',
  overlayStyle: {},
  overlayClassName: '',
  closeable: false,
  closeIconPosition: 'top-right',
  closeIcon: 'close',
  destroyOnClose: false,
  portal: null,
  overlay: true,
  round: false,
  onOpen: () => {},
  onClose: () => {},
  onOverlayClick: (e: MouseEvent) => true,
  onCloseIconClick: (e: MouseEvent) => true,
  ...defaultOverlayProps,
} as PopupProps

// 默认1000，参看variables
const _zIndex = 1100

export const Popup: FunctionComponent<
  Partial<PopupProps> & Omit<React.HTMLAttributes<HTMLDivElement>, 'title'>
> = (props) => {
  const {
    children,
    visible,
    overlay,
    closeOnOverlayClick,
    overlayStyle,
    overlayClassName,
    zIndex,
    lockScroll,
    duration,
    closeable,
    closeIconPosition,
    closeIcon,
    left,
    title,
    style,
    transition,
    round,
    position,
    className,
    destroyOnClose,
    portal,
    onOpen,
    onClose,
    onOverlayClick,
    onCloseIconClick,
    afterShow,
    afterClose,
    onClick,
  } = { ...defaultProps, ...props }
  const nodeRef = React.useRef(null)
  let innerIndex = zIndex || _zIndex
  const [index, setIndex] = useState(innerIndex)
  const [innerVisible, setInnerVisible] = useState(visible)
  const [showChildren, setShowChildren] = useState(true)
  const [transitionName, setTransitionName] = useState('')
  useLockScroll(nodeRef, innerVisible && lockScroll)
  const classPrefix = 'nut-popup'
  const baseStyle = {
    zIndex: index,
  }

  const overlayStyles = {
    ...overlayStyle,
    '--nutui-overlay-zIndex': index,
  }

  const popStyles = {
    ...style,
    ...baseStyle,
  }

  const popClassName = classNames({
    round,
    [`${classPrefix}`]: true,
    [`${classPrefix}-${position}`]: true,
    [`${className || ''}`]: true,
  })

  const closeClasses = classNames({
    [`${classPrefix}__close-icon`]: true,
    [`${classPrefix}__close-icon--${closeIconPosition}`]: true,
  })

  const open = () => {
    if (!innerVisible) {
      setInnerVisible(true)
      setIndex(++innerIndex)
    }
    if (destroyOnClose) {
      setShowChildren(true)
    }
    onOpen && onOpen()
  }

  const close = () => {
    if (innerVisible) {
      setInnerVisible(false)
      if (destroyOnClose) {
        setTimeout(() => {
          setShowChildren(false)
        }, Number(duration))
      }
      onClose && onClose()
    }
  }

  const onHandleClickOverlay = (e: MouseEvent) => {
    e.stopPropagation()
    if (closeOnOverlayClick) {
      const closed = onOverlayClick && onOverlayClick(e)
      closed && close()
    }
  }

  const onHandleClick: MouseEventHandler<HTMLDivElement> = (e: MouseEvent) => {
    onClick && onClick(e)
  }

  const onHandleClickCloseIcon: MouseEventHandler<HTMLDivElement> = (
    e: MouseEvent
  ) => {
    const closed = onCloseIconClick && onCloseIconClick(e)
    closed && close()
  }

  const onHandleOpened: EnterHandler<HTMLElement | undefined> | undefined = (
    e: HTMLElement
  ) => {
    afterShow && afterShow()
  }

  const onHandleClosed: ExitHandler<HTMLElement | undefined> | undefined = (
    e: HTMLElement
  ) => {
    afterClose && afterClose()
  }

  const resolveContainer = (getContainer: Teleport | undefined) => {
    const container =
      typeof getContainer === 'function' ? getContainer() : getContainer
    return container || document.body
  }

  const renderToContainer = (getContainer: Teleport, node: ReactElement) => {
    if (getContainer) {
      const container = resolveContainer(getContainer)
      return createPortal(node, container) as ReactPortal
    }
    return node
  }

  const renderIcon = () => {
    if (closeable) {
      return (
        <div className={closeClasses} onClick={onHandleClickCloseIcon}>
          {React.isValidElement(closeIcon) ? (
            closeIcon
          ) : (
            <Close width={12} height={12} />
          )}
        </div>
      )
    }
    return null
  }
  const renderTitle = () => {
    return (
      <>
        {position === 'bottom' && (
          <>
            {left && <div className={`${classPrefix}-left-icon`}>{left}</div>}
            {title && <div className={`${classPrefix}-title`}>{title}</div>}
          </>
        )}
      </>
    )
  }
  const renderPop = () => {
    return (
      <CSSTransition
        nodeRef={nodeRef}
        classNames={transitionName}
        unmountOnExit
        timeout={duration}
        in={innerVisible}
        onEntered={onHandleOpened}
        onExited={onHandleClosed}
      >
        <div
          ref={nodeRef}
          style={popStyles}
          className={popClassName}
          onClick={onHandleClick}
        >
          {renderTitle()}
          {renderIcon()}
          {showChildren ? children : ''}
        </div>
      </CSSTransition>
    )
  }

  const renderNode = () => {
    return (
      <>
        {overlay ? (
          <>
            <Overlay
              style={overlayStyles}
              className={overlayClassName}
              visible={innerVisible}
              closeOnOverlayClick={closeOnOverlayClick}
              zIndex={zIndex}
              lockScroll={lockScroll}
              duration={duration}
              onClick={onHandleClickOverlay}
            />
            {renderPop()}
          </>
        ) : (
          <>{renderPop()}</>
        )}
      </>
    )
  }

  useEffect(() => {
    visible && open()
    !visible && close()
  }, [visible])

  useEffect(() => {
    setTransitionName(transition || `${classPrefix}-slide-${position}`)
  }, [position, transition])

  return <>{renderToContainer(portal as Teleport, renderNode())}</>
}

Popup.defaultProps = defaultProps
Popup.displayName = 'NutPopup'
