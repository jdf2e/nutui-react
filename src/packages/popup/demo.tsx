import React, { useState } from 'react'
import { useTranslate } from '../../sites/assets/locale'
import Cell from '@/packages/cell'
import Popup from '@/packages/popup'
import './demo.scss'

interface T {
  ce5c5446: string
  c38a08ef: string
  b840c88f: string
  a74a1fd4: string
  '8dab2f66': string
  cfbdc781: string
  c3a3a1d2: string
  e51e4582: string
  '7db1a8b2': string
  a52bef0c: string
  d04fcbda: string
  '0aaad620': string
}

const PopupDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      ce5c5446: '基础类型',
      c38a08ef: '展示弹出层',
      b840c88f: '正文',
      a74a1fd4: '弹出位置',
      '8dab2f66': '顶部弹出',
      cfbdc781: '底部弹出',
      c3a3a1d2: '左侧弹出',
      e51e4582: '右侧弹出',
      '7db1a8b2': '关闭图标',
      a52bef0c: '图标位置',
      d04fcbda: '自定义图标',
      '0aaad620': '圆角弹框',
    },
    'zh-TW': {
      ce5c5446: '基礎類型',
      c38a08ef: '展示彈出層',
      b840c88f: '正文',
      a74a1fd4: '彈出位置',
      '8dab2f66': '頂部彈出',
      cfbdc781: '底部彈出',
      c3a3a1d2: '左側彈出',
      e51e4582: '右側彈出',
      '7db1a8b2': '關閉圖標',
      a52bef0c: '圖標位置',
      d04fcbda: '自定義圖標',
      '0aaad620': '圓角彈框',
    },
    'en-US': {
      ce5c5446: 'base type',
      c38a08ef: 'Show popup',
      b840c88f: 'text',
      a74a1fd4: 'popup location',
      '8dab2f66': 'top pop',
      cfbdc781: 'bottom pop',
      c3a3a1d2: 'pop up left',
      e51e4582: 'pop up right',
      '7db1a8b2': 'close icon',
      a52bef0c: 'Icon position',
      d04fcbda: 'custom icon',
      '0aaad620': 'Rounded popup',
    },
  })

  const [showBasic, setShowBasic] = useState(false)
  const [showTop, setShowTop] = useState(false)
  const [showBottom, setShowBottom] = useState(false)
  const [showLeft, setShowLeft] = useState(false)
  const [showRight, setShowRight] = useState(false)
  const [showIcon, setShowIcon] = useState(false)
  const [showIconPosition, setShowIconPosition] = useState(false)
  const [showIconDefine, setShowIconDefine] = useState(false)
  const [showBottomRound, setShowBottomRound] = useState(false)

  return (
    <>
      <div className="demo">
        <h2>{translated.ce5c5446}</h2>
        <Cell
          title={translated.c38a08ef}
          isLink
          onClick={() => {
            setShowBasic(true)
          }}
        />
        <Popup
          visible={showBasic}
          style={{ padding: '30px 50px' }}
          onClose={() => {
            setShowBasic(false)
          }}
        >
          {translated.b840c88f}
        </Popup>

        <h2>{translated.a74a1fd4}</h2>
        <Cell
          title={translated['8dab2f66']}
          isLink
          onClick={() => {
            setShowTop(true)
          }}
        />
        <Popup
          visible={showTop}
          style={{ height: '20%' }}
          position="top"
          onClose={() => {
            setShowTop(false)
          }}
        />
        <Cell
          title={translated.cfbdc781}
          isLink
          onClick={() => {
            setShowBottom(true)
          }}
        />
        <Popup
          visible={showBottom}
          style={{ height: '20%' }}
          position="bottom"
          onClose={() => {
            setShowBottom(false)
          }}
        />
        <Cell
          title={translated.c3a3a1d2}
          isLink
          onClick={() => {
            setShowLeft(true)
          }}
        />
        <Popup
          visible={showLeft}
          style={{ width: '20%', height: '100%' }}
          position="left"
          onClose={() => {
            setShowLeft(false)
          }}
        />
        <Cell
          title={translated.e51e4582}
          isLink
          onClick={() => {
            setShowRight(true)
          }}
        />
        <Popup
          visible={showRight}
          style={{ width: '20%', height: '100%' }}
          position="right"
          onClose={() => {
            setShowRight(false)
          }}
        />

        <h2>{translated['7db1a8b2']}</h2>
        <Cell
          title={translated['7db1a8b2']}
          isLink
          onClick={() => {
            setShowIcon(true)
          }}
        />
        <Popup
          closeable
          visible={showIcon}
          style={{ height: '20%' }}
          position="bottom"
          onClose={() => {
            setShowIcon(false)
          }}
        />
        <Cell
          title={translated.a52bef0c}
          isLink
          onClick={() => {
            setShowIconPosition(true)
          }}
        />
        <Popup
          closeable
          visible={showIconPosition}
          style={{ height: '20%' }}
          closeIconPosition="top-left"
          position="bottom"
          onClose={() => {
            setShowIconPosition(false)
          }}
        />
        <Cell
          title={translated.d04fcbda}
          isLink
          onClick={() => {
            setShowIconDefine(true)
          }}
        />
        <Popup
          closeable
          visible={showIconDefine}
          style={{ height: '20%' }}
          closeIcon="heart"
          position="bottom"
          onClose={() => {
            setShowIconDefine(false)
          }}
        />

        <h2>{translated['0aaad620']}</h2>
        <Cell
          title={translated['0aaad620']}
          isLink
          onClick={() => {
            setShowBottomRound(true)
          }}
        />
        <Popup
          closeable
          visible={showBottomRound}
          style={{ height: '20%' }}
          position="bottom"
          round
          onClose={() => {
            setShowBottomRound(false)
          }}
        />
      </div>
    </>
  )
}

export default PopupDemo
