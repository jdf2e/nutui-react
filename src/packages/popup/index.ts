import { Popup } from './popup'

export default Popup
