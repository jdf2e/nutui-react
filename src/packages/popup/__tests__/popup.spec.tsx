import * as React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { Popup } from '../popup'

function sleep(delay = 0): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, delay)
  })
}

test('should change z-index when using z-index prop', () => {
  const { container } = render(
    <>
      <Popup visible zIndex={99} />
    </>
  )

  const element = container.querySelector('.nut-popup') as HTMLElement
  expect(element.style.zIndex).toEqual('99')
})

test('should change animation duration when using duration prop', () => {
  const { container } = render(
    <>
      <Popup visible duration={12} />
    </>
  )

  const overlay = container.querySelector('.nut-overlay') as HTMLElement
  expect(overlay.style.animationDuration).toEqual('12s')
})

test('prop overlay-class test', async () => {
  const { container } = render(
    <>
      <Popup visible overlayClass="testclas" />
    </>
  )
  const overlay = container.querySelector('.nut-overlay') as HTMLElement
  expect(overlay).toHaveClass('testclas')
})

test('prop overlay-style test', async () => {
  const { container } = render(
    <>
      <Popup visible overlayStyle={{ color: 'red' }} />
    </>
  )
  const overlay = container.querySelector('.nut-overlay') as HTMLElement
  expect(overlay).toHaveStyle({
    color: 'red',
  })
})

test('should lock scroll when showed', async () => {
  const { rerender } = render(<Popup visible={false} />)

  rerender(<Popup visible />)

  expect(document.body.classList.contains('nut-overflow-hidden')).toBe(true)
})

test('should not render overlay when overlay prop is false', () => {
  const { container } = render(
    <>
      <Popup visible overlay={false} />
    </>
  )

  const overlay = container.querySelectorAll('.nut-overlay') as NodeListOf<Node>
  expect(overlay.length).toBe(0)
})

test('prop close-on-click-overlay test', async () => {
  const { container } = render(
    <>
      <Popup visible closeOnClickOverlay={false} />
    </>
  )
  fireEvent.click(container)
  const overlay = container.querySelector('.nut-overlay') as HTMLElement
  expect(overlay.style.display).toEqual('')
})

test('pop from top', () => {
  const { container } = render(
    <>
      <Popup visible position="top" />
    </>
  )
  const pop = container.querySelector('.popup-top') as HTMLElement
  expect(pop).toBeTruthy()
})

test('pop from bottom', () => {
  const { container } = render(
    <>
      <Popup visible position="bottom" />
    </>
  )
  const pop = container.querySelector('.popup-bottom') as HTMLElement
  expect(pop).toBeTruthy()
})

test('pop from left', () => {
  const { container } = render(
    <>
      <Popup visible position="left" />
    </>
  )
  const pop = container.querySelector('.popup-left') as HTMLElement
  expect(pop).toBeTruthy()
})

test('pop from right', () => {
  const { container } = render(
    <>
      <Popup visible position="right" />
    </>
  )
  const pop = container.querySelector('.popup-right') as HTMLElement
  expect(pop).toBeTruthy()
})

test('should render close icon when using closeable prop', () => {
  const { container } = render(
    <>
      <Popup visible closeable />
    </>
  )
  const closeIcon = container.querySelector(
    '.nut-popup__close-icon'
  ) as HTMLElement
  expect(closeIcon).toBeTruthy()
})

test('should render correct close icon when using close-icon prop', () => {
  const { container } = render(
    <>
      <Popup visible closeable closeIcon="success" />
    </>
  )

  const closeIcon = container.querySelector(
    '.nut-popup__close-icon'
  ) as HTMLElement
  expect(closeIcon.innerHTML).toMatchSnapshot()
})

test('should have "van-popup--round" class when setting the round prop', () => {
  const { container } = render(
    <>
      <Popup visible round />
    </>
  )

  const round = container.querySelector('.round') as HTMLElement
  expect(round).toBeTruthy()
})

test('should allow to using teleport prop', () => {
  render(
    <>
      <Popup visible />
    </>
  )
  expect(document.body.querySelector('.nut-popup')).toBeTruthy()
})

test('event click test', async () => {
  const { container } = render(
    <>
      <Popup visible closeOnClickOverlay />
    </>
  )
  const overlay = container.querySelector('.nut-overlay') as Element
  await fireEvent.click(overlay)
  expect(overlay).toHaveClass('hidden-render')
})

test('event click-close-icon test', () => {
  const onClickCloseIcon = jest.fn()
  const { container } = render(
    <>
      <Popup visible closeable onClickCloseIcon={onClickCloseIcon} />
    </>
  )
  const closeIcon = container.querySelector(
    '.nut-popup__close-icon'
  ) as HTMLElement
  const overlay = container.querySelector('.nut-overlay') as Element
  fireEvent.click(closeIcon)
  expect(onClickCloseIcon).toBeCalled()
  expect(overlay).toHaveClass('hidden-render')
})

test('should emit open event when prop visible is set to true', () => {
  const onOpen = jest.fn()
  const { rerender } = render(
    <>
      <Popup visible={false} onOpen={onOpen} />
    </>
  )

  rerender(
    <>
      <Popup visible onOpen={onOpen} />
    </>
  )

  expect(onOpen).toBeCalled()
})

test('event click-overlay test', async () => {
  const onClickOverlay = jest.fn()
  const { container } = render(
    <>
      <Popup visible onClickOverlay={onClickOverlay} />
    </>
  )

  const overlay = container.querySelector('.nut-overlay') as Element
  fireEvent.click(overlay)
  expect(onClickOverlay).toBeCalled()
})
