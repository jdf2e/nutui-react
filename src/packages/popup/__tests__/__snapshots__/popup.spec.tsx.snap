// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correct close icon when using close-icon prop 1`] = `"<i class=\\" nutui-iconfont nut-icon nutui-icon nut-icon-success \\" style=\\"font-size: 12px; width: 12px; height: 12px;\\"></i>"`;
